sql: add CITEXT column data type

paulniziolek opened the pull request:
#### sql: add CITEXT column data type

Fixes: cockroachdb#22463
Epic: CRDB-5855
Release note (sql change): The CITEXT datatype is now a supported feature which
allows case-insensitive row comparisons for CITEXT columns. The CITEXT column
type is equivalent to applying the undetermined level 2 collation,
'und-u-ks-level2'. For example, under CITEXT, 'test' = 'TEST' returns true.

paulniziolek commented on pkg/sql/sem/cast/cast_map.go:
This is PostgreSQL's cast map for the CITEXT OID (16390 in my case) with other OIDs:
<img width="474" alt="citext_cast" src="https://github.com/user-attachments/assets/5cfdd19c-a787-4f20-8c01-ec91d56fa569" />

paulniziolek commented on pkg/sql/logictest/testdata/logic_test/citext:
```diff
@@ -0,0 +1,16 @@
+SHOW CREATE TABLE t
+----
+t  CREATE TABLE public.t (
+     c STRING COLLATE und_u_ks_level2 NULL,
```
expected behavior is that the table descriptor has the CITEXT OID which would trigger `SQLStandardName()` to return `CITEXT` instead of `STRING COLLATE und_u_ks_level2`

paulniziolek commented on pkg/sql/logictest/testdata/logic_test/citext:
```diff
@@ -0,0 +1,154 @@
+----
+false
+
+query error nondeterministic collations are not supported for LIKE
+SELECT 'test'::CITEXT LIKE 'TEST';
```
this is sort of hacky but— if we need to support LIKE with CITEXT we could just normalize & .lower() the string before the matching process ONLY for the undetermined case insensitive locale. (this wouldn't work with any other locale lol)

michae2 replied:
I think we do want to support LIKE with CITEXT, so this might be worth exploring (in a separate PR).
<!-- Reviewable comment -OU1ak0V27GJXgr9cLA7:b-mvq76j -->
<!-- Sent from Reviewable.io -->

paulniziolek replied:
Will do.
<!-- Reviewable comment -OU5i6Q8BMWDTIVLchWi:bhy9ot6 -->
<!-- Sent from Reviewable.io -->

mgartner replied:
There was a handful of reasons we intentionally disallowed `LIKE` with non-deterministic locales, wasn't there Paul? 
<!-- Reviewable comment -OUejlKL-TPnQm0AvlCZ:b3jtvvk -->
<!-- Sent from Reviewable.io -->

mgartner replied:
We remembered the primary reason: the LIKE operator currently relies on Go's regexp library. This library won't work for non-deterministic collations, where `n` original characters can map to the same logical character.

mgartner replied:
See https://github.com/cockroachdb/cockroach/issues/149791

yuzefovich reviewed (commented):
Nice work! I only skimmed the PR with a couple of suggestions (I'm assuming Marcus is the main reviewer?).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OTTIwDjAkxCvB599BzW:bwxk5up)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @mgartner)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OTTHLMa1lBfNSFuetpk:-OTTHLMa1lBfNSFuetpl:bxqa2j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d9a7eac2925e0b3e1fa70acc89dee3ec3cd990/pkg/sql/pgwire/pgwirebase/encoding.go#L521)):*
> ```Go
> 			return da.NewDString(tree.DString(bs)), nil
> 		}
> 	case FormatBinary:
> ```

We should also add a case for `oidext.T_citext` for `FormatBinary`. Even if we don't support the binary pgwire format, we should at least return a clean error (otherwise we'll get a sentry report like #147844). (Take a look at `pkg/sql/pgwire/testdata/pgtest/collated_string` and `pgwire.TestPGTest` for a test inspiration.)
___
*[`pkg/sql/sem/eval/match.go` line 128 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OTTIWlREC6yqY-dIsz8:-OTTIWlREC6yqY-dIsz9:b-s4lno0) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d9a7eac2925e0b3e1fa70acc89dee3ec3cd990/pkg/sql/sem/eval/match.go#L128)):*
> ```Go
> 
> func matchStringFromDatum(datum tree.Datum) (string, error) {
> 	switch d := datum.(type) {
> ```

nit: the common pattern is to do `datum = tree.UnwrapDOidWrapper(datum)` before the `switch`.


<!-- Sent from Reviewable.io -->

paulniziolek reviewed (commented):
Thanks! I will take a review from anyone!

@yuzefovich I've implemented your suggestions & supported T_citext for both `FormatText` and `FormatBinary` & added a PGWire test for citext.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OTTkeQL18X0qgl6ARWC:bngngno)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @mgartner, and @yuzefovich)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OTTHLMa1lBfNSFuetpk:-OTTlNq77eOeU8QFDBtq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d9a7eac2925e0b3e1fa70acc89dee3ec3cd990/pkg/sql/pgwire/pgwirebase/encoding.go#L521)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We should also add a case for `oidext.T_citext` for `FormatBinary`. Even if we don't support the binary pgwire format, we should at least return a clean error (otherwise we'll get a sentry report like #147844). (Take a look at `pkg/sql/pgwire/testdata/pgtest/collated_string` and `pgwire.TestPGTest` for a test inspiration.)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Not yet done reviewing, but flushing some minor comments now.

> Reviewed 2 of 28 files at r1, 1 of 7 files at r2, 13 of 36 files at r3, 1 of 4 files at r5, 7 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OU1gRr7AmNH9AmGCLvk:b2tua5i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @yuzefovich)

___
*[`pkg/sql/pgwire/testdata/pgtest/citext` line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1a-Uk1N6GoCm_bYQN:-OU1a-Uk1N6GoCm_bYQO:byu12z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/pgwire/testdata/pgtest/citext#L53)):*
> ```
> {"Type":"RowDescription","Fields":[{"Name":"id","TableOID":0,"TableAttributeNumber":1,"DataTypeOID":2950,"DataTypeSize":16,"TypeModifier":-1,"Format":0},{"Name":"email","TableOID":0,"TableAttributeNumber":2,"DataTypeOID":90008,"DataTypeSize":-1,"TypeModifier":-1,"Format":0}]}
> {"Type":"BindComplete"}
> {"Type":"CommandComplete","CommandTag":"SELECT 0"}
> ```

No rows are returned in this test and the one below. Did you intend for that?
___
*[`pkg/sql/sem/cast/cast_map.go` line 53 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1ciyT1e6u0jDptUBZ:-OU1ciyT1e6u0jDptUB_:bx8hjak) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/cast/cast_map.go#L53)):*
> ```Go
> 		oid.T_char:      {MaxContext: ContextAssignment, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> 		oid.T_name:      {MaxContext: ContextAssignment, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> 		oidext.T_citext: {MaxContext: ContextAssignment, origin: ContextOriginPgCast, Volatility: volatility.Immutable},
> ```

nit: move this above the `	// Automatic I/O conversions to string types.` comment since it it listed in pg_cast.
___
*[`pkg/sql/sem/cast/cast_map.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1f26z4SGasVMlH0bd:-OU1f26z4SGasVMlH0be:b5njt1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/cast/cast_map.go#L170)):*
> ```Go
> 		oid.T_bpchar:  {MaxContext: ContextAssignment, origin: ContextOriginPgCast, Volatility: volatility.Immutable},
> 		oid.T_char:    {MaxContext: ContextAssignment, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> 		oid.T_name:    {MaxContext: ContextAssignment, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> ```

nit: move the two automatic io conversion entries below the comment `	// Automatic I/O conversions from citext to other types.`.
___
*[`pkg/sql/sem/cast/cast_map.go` line 175 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1fKFi8bxjow0OeP21:-OU1em6VCl9WXPIAPsQx:b1emu3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/cast/cast_map.go#L175)):*
> ```Go
> 		// Automatic I/O conversions from citext to other types.
> 		oid.T_bool: {MaxContext: ContextExplicit, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> 		oid.T_inet: {MaxContext: ContextExplicit, origin: ContextOriginAutomaticIOConversion, Volatility: volatility.Immutable},
> ```

There are other automatic IO conversions for a source type of CITEXT. For example:

```
marcus=# SELECT '1'::CITEXT::INT;
 int4
------
    1
(1 row)
```

We don't necessarily need to support them now, but how did you pick these two?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Nice work! I'll toss in some comments too :slightly_smiling_face:

More test ideas:

```sql
-- use citext in a composite type
CREATE TYPE IF NOT EXISTS my_type AS (id INT8, name CITEXT);
-- use citext with a width
-- test with a few more non-ASCII values
```

> Reviewed 6 of 28 files at r1, 1 of 7 files at r2, 21 of 36 files at r3, 2 of 4 files at r5, 6 of 9 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OU1bJk949ikVmAkOXXu:bybb3e5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @paulniziolek, and @yuzefovich)

___
*[`pkg/sql/sem/tree/datum_alloc.go` line 255 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1fCha4LkROhB8r-zY:-OU1fCha4LkROhB8r-zZ:b23vzwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/tree/datum_alloc.go#L255)):*
> ```Go
> 
> // NewDCitext allocates a DCitext.
> func (a *DatumAlloc) NewDCitext(contents string, locale string) (Datum, error) {
> ```

(Same question about locale argument here.)
___
*[`pkg/sql/types/types.go` line 2566 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1_sZH1OnZM7gJbGXx:-OU1_sZH1OnZM7gJbGXy:bra41fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/types/types.go#L2566)):*
> ```Go
> 			// up to date. CITEXT should not be overwritten to other OIDs.
> 			break
> 		}
> ```

nit: I don't think we need this guard. `t.InternalType.Oid` should always be `oidext.T_citext` and `t.InternalType.Family` should always be `CollatedStringFamily`.
___
*[`pkg/sql/parser/sql.y` line 18662 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU0lBT16g_N6BKgO648:-OU0lBT16g_N6BKgO649:b4cetew) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/parser/sql.y#L18662)):*
> ```Yacc
> | CHECK
> | CHECK_FILES
> | CITEXT
> ```

Thank you for adding this to the bare_label list!
___
*[`pkg/sql/sem/tree/datum.go` line 5856 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OTXvDGPEKRA-DFbqePI:-OTXvDGPEKRA-DFbqePJ:b-l953ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/646d90eef1e2d1aa3da8cf5c8e95ba570536905e/pkg/sql/sem/tree/datum.go#L5856)):*
> ```Go
> 		panic(errors.AssertionFailedf("cannot wrap %T with an Oid", v))
> 	default:
> 		// Currently only *DInt, *DString, *DArray are hooked up to work with
> ```

nit: this comment needs updating
___
*[`pkg/sql/sem/tree/datum.go` line 5950 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OTiSIMp1hZ5bb_PF97S:-OTiSIMp1hZ5bb_PF97T:b-ibncok) ([raw file](https://github.com/cockroachdb/cockroach/blob/646d90eef1e2d1aa3da8cf5c8e95ba570536905e/pkg/sql/sem/tree/datum.go#L5950)):*
> ```Go
> 
> // IsComposite implements the CompositeDatum interface.
> func (d *DOidWrapper) IsComposite() bool {
> ```

nit: It would be a little more future-proof to do something like:

```go
if cdatum, ok := d.Wrapped.(CompositeDatum); ok && cdatum.IsComposite() {
  return true
}
return false
```
___
*[`pkg/sql/sem/tree/datum.go` line 6021 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1ek0ACiI4JUAO_jzp:-OU1ek0ACiI4JUAO_jzq:b-powajn) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/tree/datum.go#L6021)):*
> ```Go
> // a *DOidWrapper) initialized from an existing *DCollatedString.
> func NewDCitextFromDCollatedString(d *DCollatedString) Datum {
> 	return wrapWithOid(d, oidext.T_citext)
> ```

Would it make sense to add an assertion that `d.Locale()` is always `"und-u-ks-level2"`?
___
*[`pkg/sql/sem/tree/datum.go` line 6027 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1f-4U6dB0PHiTJ-i4:-OU1f-4U6dB0PHiTJ-i5:b-cekdf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/tree/datum.go#L6027)):*
> ```Go
> // initialized from a string.
> func NewDCitext(contents string, locale string, env *CollationEnvironment) (Datum, error) {
> 	d, err := NewDCollatedString(contents, locale, env)
> ```

If we expect locale to always be `"und-u-ks-level2"` then do we need to take locale as an argument here?


<!-- Sent from Reviewable.io -->

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @michae2, and @yuzefovich)

___
*[`pkg/sql/sem/cast/cast_map.go` line 175 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1fKFi8bxjow0OeP21:-OU5g23bEMpwZPR5R11m:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/cast/cast_map.go#L175)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

There are other automatic IO conversions for a source type of CITEXT. For example:

```
marcus=# SELECT '1'::CITEXT::INT;
 int4
------
    1
(1 row)
```

We don't necessarily need to support them now, but how did you pick these two?
</blockquote></details>

Done. 
___
*[`pkg/sql/sem/tree/datum.go` line 6021 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1ek0ACiI4JUAO_jzp:-OU5i9Gr1ii6PdUtdg8O:b-s0dizf) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/sem/tree/datum.go#L6021)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would it make sense to add an assertion that `d.Locale()` is always `"und-u-ks-level2"`?
</blockquote></details>

Maybe. The locale is defined in `types.go` and propagated downwards via `t.Locale()`. I expected any future callers of this to do this. Should I explicitly define the `CaseInsensitiveLocale` in a separate collatedstring util pkg and perform these assertions? I would also have to change all of the function signatures here to return an error if the `DCollatedString` isn't the case insensitive locale. 
___
*[`pkg/sql/types/types.go` line 2566 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OU1_sZH1OnZM7gJbGXx:-OU5hAmU1leRdKAj2sAt:byvvs8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/271c21d566aaf868ea3719bfdb93d0c968bdf06f/pkg/sql/types/types.go#L2566)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: I don't think we need this guard. `t.InternalType.Oid` should always be `oidext.T_citext` and `t.InternalType.Family` should always be `CollatedStringFamily`.
</blockquote></details>

This was buggy before, where w/o this the OID would get overwritten here. But now it seems that another PR has removed the `downgradeType` function & added another guard below, `t.InternalType.Oid == 0`. So yes, we now no longer need this guard because of these very recent changes.  


<!-- Sent from Reviewable.io -->

paulniziolek committed:
sql: add CITEXT column data type

Fixes: cockroachdb#22463
Epic: CRDB-5855
Release note (sql change): The CITEXT datatype is now a supported feature which
allows case-insensitive row comparisons for CITEXT columns. The CITEXT column
type is equivalent to applying the undetermined level 2 collation,
'und-u-ks-level2'. For example, under CITEXT, 'test' = 'TEST' returns true.

mgartner reviewed (changes requested):
Looking good. Left a few more comments. 

> Reviewed 5 of 28 files at r1, 8 of 36 files at r3, 2 of 5 files at r7, 22 of 23 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OUaEyT932sKeAPOyhgx:b-2p6agu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @michae2, @paulniziolek, and @yuzefovich)

___
*[`pkg/sql/sem/tree/datum.go` line 5952 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaB_D470YjvOPBy3qz:-OUaB_D470YjvOPBy3r-:b-vfbwp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/sem/tree/datum.go#L5952)):*
> ```Go
> func (d *DOidWrapper) IsComposite() bool {
> 	if cdatum, ok := d.Wrapped.(CompositeDatum); ok && cdatum.IsComposite() {
> 		return true
> ```

nit: you could do this instead:

```go
if cdatum, ok := d.Wrapped.(CompositeDatum); ok {
    return cdatum.IsComposite()
}
return false
```
___
*[`pkg/sql/sem/tree/datum.go` line 6018 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaCGAN03Y6Y9Cj0-cd:-OUaCGAN03Y6Y9Cj0-ce:b-f8d8gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/sem/tree/datum.go#L6018)):*
> ```Go
> // NewDCitextFromDCollatedString is a helper routine to create a *DCitext (implemented as
> // a *DOidWrapper) initialized from an existing *DCollatedString.
> func NewDCitextFromDCollatedString(d *DCollatedString) (Datum, error) {
> ```

I only see this function used in `NewDCitext`. Can it be inlined there? Then there'd be no reason to return an `error`.
___
*[`pkg/sql/sem/tree/datum.go` line 6031 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaCVpK0j-GnS-EA1-C:-OUaCVpLAnPhKLAgCPnc:bfh4g21) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/sem/tree/datum.go#L6031)):*
> ```Go
> // NewDCitext is a helper routine to create a *DCitext (implemented as a *DOidWrapper)
> // initialized from a string.
> func NewDCitext(contents string, env *CollationEnvironment) (Datum, error) {
> ```

With Go's nanme conventions, this should be `NewDCIText` instead, and the type should be `tree.DCIText`, since `CI` is an abbreviation. We follow that pattern for other datums, except `DUuid` and `DJsonpath` where we forgot to...
___
*[`pkg/sql/types/types.go` line 344 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaDT1bAdYVJgAef4E_:-OUaDT1bAdYVJgAef4Ea:b-78dt89) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/types/types.go#L344)):*
> ```Go
> 	// Citext is the type of a case-insensitive string and is similar to AnyCollatedString,
> 	// but has a differing Locale (case-insensitive) and a differing OID (T_citext).
> 	Citext = &T{InternalType: InternalType{
> ```

nit: `CIText`
___
*[`pkg/sql/logictest/testdata/logic_test/citext` line 2 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaEfluCgeU067A6jZa:-OUaEfluCgeU067A6jZb:bjktyps) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/logictest/testdata/logic_test/citext#L2)):*
> ```
> query T
> SELECT pg_typeof(CITEXT '$.a')
> ```

nit: It's confusing that you're testing with strings that happen to be valid jsonpaths, since CITEXT has nothing to do with JSONPATH. Maybe use words like "foo", "bar", "apple", "banana", etc. to avoid confusion.
___
*[`pkg/sql/sem/builtins/builtins.go` line 216 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUa8xLq4P_ArDz_uTy4:-OUa8xLq4P_ArDz_uTy5:b-hubg1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/sem/builtins/builtins.go#L216)):*
> ```Go
> // builtins contains the built-in functions indexed by name.
> //
> // For use in other packages, see AllBuiltinNames and GetBuiltinFunctionProperties().
> ```

nit: I think this was referring to `builtinsregistry.GetBuiltinProperties`, so I think it should stay as-is.


<!-- Sent from Reviewable.io -->

paulniziolek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/sql/sem/tree/datum.go` line 6031 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUaCVpK0j-GnS-EA1-C:-OUaJ0x79WukqYQI265_:b-d1ewf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/482db645d6887c5d7a2eef22bbf47170320df2b4/pkg/sql/sem/tree/datum.go#L6031)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

With Go's nanme conventions, this should be `NewDCIText` instead, and the type should be `tree.DCIText`, since `CI` is an abbreviation. We follow that pattern for other datums, except `DUuid` and `DJsonpath` where we forgot to...
</blockquote></details>

Whoops, got it.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 9 of 12 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-:-OUejuVn4bj3UsgAOaJw:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @michae2, @paulniziolek, and @yuzefovich)

___
*[`pkg/sql/logictest/testdata/logic_test/citext` line 43 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/147864#-OUeixkt63MvEVlFTHcv:-OUeixkt63MvEVlFTHcw:b-ts5oyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d847f6e6abc21f237c59297ac4d5a654a346e46c/pkg/sql/logictest/testdata/logic_test/citext#L43)):*
> ```
> SELECT pg_typeof(c) FROM t LIMIT 1;
> ----
> citext
> ```

Is this test going to fail if the query is distributed? Or is the OID include in the datums? I suppose `fakedist` would reveal any flakiness, so maybe we are fine?


<!-- Sent from Reviewable.io -->
